### PR TITLE
feat(x/evm): Token Middleware

### PR DESCRIPTION
**Pull request by itsdevbear**

### CODE REVIEW

File: cosmos/x/erc20/keeper/middleware.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	// Get the ERC20 token corresponding to a given denom.
+	if _, err := k.DenomKVStore(ctx).GetAddressForDenom(denom); err != nil {
+		// If the denom is found, we need to burn the corresponding bank denom.
+		if err = k.BurnCoinsFromAddress(ctx, sender, denom, amount); err != nil {
+			return false, err
+		}
+		// If the token was found, that means we don't have to deploy it. So we need to unlock the lock.
+		k.deployLock.Unlock()
```
Comment (itsdevbear) [inline cosmos/x/erc20/keeper/middleware.go:75]: Defer above if, possible deadlock

**Comment by itsdevbear** [comment]
The more I thing on this, this is such an ugly solution, we should just bite the bullet and do the Create and Call from the precompile. 

I think it's important to show it can be done in an safe way.

**Comment by itsdevbear** [comment]
Having the smart contract perform logic goes against the ethos of why we even built Polaris in the first place.

**Comment by itsdevbear** [comment]
though @smsunarto I'd love your thoughts here. 

Right now we don't allow precompile to call back into the EVM. This means that the ERC20 precompile can't call transfer or transferFrom on a token to move it from someone's wallet after they approve the ERC20 module address, so we have to do it this way.

**Comment by itsdevbear** [comment]
If we were to allow it, the precompile could just call CREATE or CREATE2 to deploy the PolarisERC20 and then CALL to transfer, mint and burn tokens. 

Which feels more correct, though slightly more complex. 

The shim contract feels gross to me.

File: contracts/src/cosmos/x/erc20/ModuleRouter.sol [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+    }
+
+    /**
+     * @dev Transfer tokens to Cosmos
+     * @param denom The denom to transfer.
+     * @param recipient The address of the recipient
+     * @param amount The amount of tokens to transfer
+     */
+    function transferToCosmos(string memory denom, address recipient, uint256 amount) public {
```
Comment (smsunarto) [inline contracts/src/cosmos/x/erc20/ModuleRouter.sol:75]: nit, lets make the ordering consistent.

`denom, amount, recipient`

File: contracts/src/cosmos/x/erc20/ModuleRouter.sol [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+     * @param _erc20Module The address of the ERC20Module
+     */
+    constructor(IERC20Module _erc20Module) {
+        erc20Module = _erc20Module;
+    }
+
+    /**
+     * @dev Transfer tokens to Cosmos
+     * @param token The address of the token to transfer
+     * @param amount The amount of tokens to transfer
```
Comment (smsunarto) [inline contracts/src/cosmos/x/erc20/ModuleRouter.sol:52]: nit, the position of amount/recipient is switched.

**Comment by itsdevbear** [comment]
@calbera we gonna nuke the solidity ya?

**Comment by calbera** [comment]
yee

**Comment by itsdevbear** [comment]
closing in favour of: https://github.com/berachain/polaris/pull/524

**Pull request closed without merging**